Improve: Don't spellcheck 1 or 2 letter words

Kebap opened the pull request:
#### Brief overview of PR changes/additions
Only start spellchecking on words with a length of at least 3.

Before, in 4.16.0 release for example:
![image](https://user-images.githubusercontent.com/117238/219877479-e78b6b87-e329-401f-836d-24c15bd950dc.png)

Now:
![image](https://user-images.githubusercontent.com/117238/219880355-cce479d2-95aa-4c49-aa12-53369fd45059.png)

But:
![image](https://user-images.githubusercontent.com/117238/219880376-083df31a-7a3f-4bc5-b91e-863281d33c40.png)

#### Motivation for adding to Mudlet
More intuitive user experience.

#### Other info (issues closed, discussion etc)
Fix #2873
Also refactor / extract from ultra long function a tiny bit at a time:
- new `TCommandLine::fillSpellCheckList(QMouseEvent* event, QMenu* popup)`
- extracted from `void TCommandLine::mousePressEvent(QMouseEvent* event)`

Kebap committed:
reverse logic

Kebap committed:
Typo

Kebap committed:
Undo

Kebap committed:
review string size before spellcheck

Kebap committed:
extract spellcheck listing

Kebap committed:
register new function in header

Kebap committed:
actually can simply return now

Kebap committed:
fix whitespace in comments (important!)

Kebap committed:
reduce variable length

Kebap committed:
use cpp17 style if init-statement

Kebap committed:
do you copy?

Kebap committed:
yeah the comments are all over the place

SlySven commented:
~~Awaiting a decision on: https://github.com/Kebap/Mudlet/pull/11.~~ *Decided that it was a :-1:*

vadi2 reviewed (commented):
This still spellchecks accented characters incorrectly and @SlySven's PR fixes that. As I mentioned in https://discord.com/channels/283581582550237184/283582439002210305/1077129447469224006 I think his PR was fine (actually - it was good), just the UI setting was not necessary.

![image](https://user-images.githubusercontent.com/110988/220296359-fa25b29f-2ca0-4eab-aa56-a4f84f1d2332.png)

SlySven commented:
> This still spellchecks accented characters incorrectly and @SlySven's PR fixes that. As I mentioned in https://discord.com/channels/283581582550237184/283582439002210305/1077129447469224006 I think his PR was fine (actually - it was good), just the UI setting was not necessary.
> 
> ![image](https://user-images.githubusercontent.com/110988/220296359-fa25b29f-2ca0-4eab-aa56-a4f84f1d2332.png)

Are those done with combining diacriticals (U+0061 {LATIN SMALL LETTER A} + U+0301 {COMBINING ACUTE ACCENT}) or the single accented characters (U+00E1 {LATIN SMALL LETTER A WITH ACUTE})?

We don't, for reasons of code simplicity, "Normalise" incoming data, so we have to handle both inputs {*normalising* the former would convert that particular case to the latter - but it would also change the number of `QChar`s used to represent the grapheme - which would put the `QChar` and `TChar` for that line out of sync.}

To fix that, so that we work on the grapheme rather than `QChar` count in the content you'd need to change:
```diff
-     if (bool IsWordLongEnoughForSpellCheck = mSpellCheckedWord.size() > 2; !IsWordLongEnoughForSpellCheck) {
+     if (bool IsWordLongEnoughForSpellCheck = TBuffer::::lengthInGraphemes(mSpellCheckedWord) > 2; !IsWordLongEnoughForSpellCheck) {
          return;
      }
```

For future developments I would strongly urge you to change that `2` though to an integer type class member {I used `(int) mSpellCheckedMinLength`}. This is particularly as I think there is a second place where it is needed that is not in this PR currently - in the `(void) TCommandLine::spellCheckWord(QTextCursor&)` method.
```diff
     QTextCharFormat f;
     mSpellChecking = true;
     c.select(QTextCursor::WordUnderCursor);
+    if (bool IsWordLongEnoughForSpellCheck = (TBuffer::lengthInGraphemes(c.selectedText()) >= mSpellCheckedMinLength); !IsWordLongEnoughForSpellCheck) {
+        // We are bailing out because the word is not now long enough, just
+        // in case mSpellCheckedMinLength increased the minimum length
+        // since the last time we did this word we had better undo any prior
+        // underline, and we need to also reset the flag:
+        f.setFontUnderline(false);
+        c.setCharFormat(f);
+        setTextCursor(c);
+        mSpellChecking = false;
+        return;
+    }
+
     QByteArray encodedText = mpHost->mpConsole->getHunspellCodec_system()->fromUnicode(c.selectedText());

```

Kebap committed:
cpp17 style if-init was too unwieldy here

Kebap committed:
measure length of graphemes properly

Kebap committed:
store min length for future reference

Kebap committed:
make new variable sound nicer

Kebap committed:
add more feedback from SlySven

Kebap committed:
drop unused variable

Kebap commented:
> I think his PR was fine (actually - it was good), just the UI setting was not necessary.

Do you need the minimum length be exported / imported with profile as well?

Kebap committed:
whitespaces

vadi2 commented:
I don't see this option being changed frequently by users, only by developers - so an application-wide setting will be enough.

An example is the profile autosave (`autosaveIntervalMinutes`) - no users needed to change it - but it's there as  for any kind of development work.

Kebap committed:
Use >= over > here

Kebap commented:
Not sure if I understand correctly. Will it be sufficient to make the minimum value available as class member, like SlySven outlined above, or do you need me to look into QSettings like autosaveIntervalMinutes does?

Usability-wise, I am not sure why especially developers would want to change the minimum length of words from when their spell-check happens, or why they would like to do it more than regular players.

vadi2 commented:
> Usability-wise, I am not sure why especially developers would want to change the minimum length of words from when their spell-check happens, or why they would like to do it more than regular players.

Just in case they're fixing something. The autosave interval for profiles doesn't need to be edited much at all, but when you're trying to do some work related to saves (such as https://github.com/Mudlet/Mudlet/issues/6414), then it can be handy to disable the option. 

So it's a 'once in a blue moon' kind of thing.

Kebap commented:
I guess that means I need to learn about QSettings and the current draft is not sufficient.

Kebap committed:
Merge branch 'development' into spellcheck_2_characters

vadi2 commented:
Try this - move this setting to Host, and see how `autosaveIntervalMinutes` is loaded and saved.

Kebap committed:
Move min length to Host

Kebap committed:
make it public

vadi2 committed:
Merge branch 'development' into spellcheck_2_characters

vadi2 committed:
Add ability to set spell check minimum length and map autosave interval in Host class

vadi2 committed:
Migrate spellcheck value to a single one in mudlet class

vadi2 commented:
https://github.com/Mudlet/Mudlet/pull/6553/commits/40a1360bd29e2b25885cee8fd91af4e3f592396b added the on/off switch for the spellchecking and https://github.com/Mudlet/Mudlet/pull/6553/commits/341f6fc124146cf5f397767142ffd3718f96849e moved it to the `mudlet` class since this value was getting duplicated across Host classes (and that's a bit messy since it really is just a global value). With this change it looks good to me!

vadi2 committed:
Merge branch 'development' into spellcheck_2_characters

vadi2 commented:
Resolved conflicts and updated to latest `development`.

SlySven reviewed (commented):
What has the map autosave got to do with spellchecking - this is the sort of scope creep I get accused of! :worried: 

***Edit: this comment was made without a consideration of the wider context - that this is internal and not user visible...!***

SlySven reviewed (dismissed):
:-1: As spelling (and the language to use for it is very much a per-profile setting) it is NOT a good idea to move settings for it from the per-profile ones to a centralised one.

If you are bi/multi-lingual - and I know that applies to both of you @Kebap and @vadi2 - do you always play MUDs that ALL use the SAME language?

Please rethink this.

vadi2 commented:
Nothing, but it's not scope creep either - nothing with the map autosave got changed at all. I was just going to introduce a problem: settings being loaded twice in the same class - so I avoided the problem by keeping them to be loaded just once.

Where did I mention scope screep on a similar nature?

vadi2 commented:
> If you are bi/multi-lingual - and I know that applies to both of you @Kebap and @vadi2 - do you always play MUDs that ALL use the SAME language?

This is just a simple on/off switch for a developer or someone debugging. This is not intended to be a user-settable option that they'll need to be fiddling with often. It's either this, or no option at all.

We're overflowing with random options that no players are ever going to change in the settings window already 😅

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -672,192 +672,202 @@ void TCommandLine::slot_popupMenu()
+#endif
+        if (mpHost->mEnableSpellCheck) {
+            fillSpellCheckList(event, popup);
+            // else the word is in the dictionary - in either case show the context
```
:confused:  This comment doesn't make sense in this position now...

Kebap replied:
Would you suggest it makes more sense to move its position to the end of TCommandLine::fillSpellCheckList() then?

Then again I did not change its position at all, so maybe the original position did not make the most sense either?

SlySven commented on src/TCommandLine.cpp:
```diff
@@ -672,192 +672,202 @@ void TCommandLine::slot_popupMenu()
+    * Remove word action
+    * --------- separator_aboveStandardMenu
+    *
+    * The insertAction[s](...)/(Separator(...)) insert their things
+    * second argument (or generated by themself) before the first (or
+    * only) argument given.
```
This might clarify it a bit more - I recently revisited this area (to add controls for the PR for saving the command line history #6767) and it took me a while to get my head around the code that builds the context menu... :astonished: 
```suggestion
    * The insertAction[s](...)/(Separator(...)) insert their things
    * second argument (or generated by themself) before the first (or
    * only) argument given; they return a pointer to the item they
    * and we need that to pass into the next one we create - which is
    * why we are repeatedly capturing those pointers into `auto`
    * typed variables.
```

Kebap replied:
I don't agree that longer is always clearer, also you introduce weird phrasing like:
"they return a pointer to the item they and we need that"

```suggestion
    * The following "insertAction", "insertActions" and "insertSeparator" need as
    * their first argument a pointer to the item before which they should insert.
    * They also return a pointer to the item they.
    * Thus we capture those pointers to give to the next functions repeatedly.
```

SlySven commented on src/Host.cpp:
```diff
@@ -475,10 +477,9 @@ void Host::loadMap()
 {
-    auto settings = mudlet::self()->getQSettings();
-    if (auto interval = settings->value("autosaveIntervalMinutes", 2).toInt(); interval > 0) {
+    if (interval > 0) {
```
Should there not be code to handle the `<1` case to run `killTimer(id)` on the (greater than 0) value that the `(int) QObject::startTimer(int interval, Qt:TimerType = Qt::CoarseTimer)` call returns?

vadi2 replied:
Not really, this function is only ever called once on Mudlet start

SlySven replied:
Yeah, but if we ever get into a situation where we want to debug this during a run there is no way to do that as we are not capturing the returned value - so we cannot stop it with the current implementation in any way after it has been started...

vadi2 replied:
I don't think we need to build extensive debugging facilities for this, this code did not need debugging often - and when it will need it, well it's a 2min job to build it. Also this would definitely be scope creep as it's not solving any problem related to spellchecking 😅

SlySven reviewed (commented):
I think we need to document (in the relevant header file) that the `mMinLengthForSpellCheck` and if there were a `(int) mudlet::mAutoSaveInterval; // minutes` member that they are internal (debugging) members that are not anticipated to be end-user accessible...

Kebap committed:
Merge branch 'development' into spellcheck_2_characters

Kebap commented:
Thanks to Vadi for adding the capability to influence the minimum word length by yourself now.

I will mark this as ready for review then once again.

vadi2 commented:
Looks good to me, merge whenever you'd like!

Kebap committed:
Merge branch 'development' into spellcheck_2_characters

SlySven committed:
Merge branch 'development' into spellcheck_2_characters

Conflicts resolved in:
* src/TCommandLine.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: const-ify some bits of the main PR

In accordance with #6843 which has been applied to the 'development' branch
code since the main PR was started.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Merging this into `development` is now non-trivial - but I think I have been able to in https://github.com/Kebap/Mudlet/pull/12 if that can be tested by you @kebap or @vadi2 and then applied to this PR...

SlySven committed:
Revise: include a bit missed from last commit!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap committed:
Merge branch 'development' into spellcheck_2_characters, also merge pull request #12 from SlySven/spellcheck_2_characters

Merge plus a few changes, see PR description and comments for details

Kebap commented:
After merge, this is now much easier to review what happened in your branch then before the merge.
Still my findigs commented in [there](https://github.com/Kebap/Mudlet/pull/12#issuecomment-1609169810) seem to hold true.


@SlySven Not sure how your changes affect things, as I hardly deal with Unicode/UTF-8 characters, and/or spellchecks.

@vadi2 Does your approval of merging this still hold true after the recent changes introduced, or do you want to reconsider?

vadi2 commented:
I'm good!

SlySven commented:
:hankey: Hold it - I messed up - I forgot that though the per profile/shared dictionary is always UTF-8 encoded (keeps things simple) the system ones might not be. That is why there was the `(QByteArray) encodedText` variable - it holds the word being spell-checked/alternate spelling suggested for in the encoding of the system dictionaries. So my changes to use it for the per profile/shared one was rong.

I need to undo that - and comment it to avoid that again...

***Update:***
Okay - the fix for that is mentioned under here now.

As for the other thing that @Kebap mentioned about removing the braces confining the scope for the "is the word long enough to bother doing spelling checks" - they are unneeded - if the test fails the rest of the method is skipped anyway and if the test passes the only thing that gets added to the method scope as a whole is the boolean flag about whether to do the work - and that isn't used anywhere else in the method so it does no harm to leave it in scope for the rest of the method.

SlySven committed:
Fix: undo mistake in prior PR

I forgot that the per-profile/shared dictionary is always UTF-8 encoded
whereas the "system" one may not be, so the word that the spelling checks/
suggestions are done on must be passed to the `hunspell` library as UTF-8
encoded bytes for the former but as the appropriate encoding for the
latter; it was an error to use the encoded bytes for the latter for the
former which I mistakenly thought was a simplification/improvement I could
make in https://github.com/Kebap/Mudlet/pull/12 !

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> As for the other thing that @Kebap mentioned about removing the braces confining the scope for the "is the word long enough to bother doing spelling checks" - they are unneeded - if the test fails the rest of the method is skipped anyway and if the test passes the only thing that gets added to the method scope as a whole is the boolean flag about whether to do the work - and that isn't used anywhere else in the method so it does no harm to leave it in scope for the rest of the method.

It's a matter of taste at this point, that is why I reluctantly agreed to your change.
I agree, it does no harm (yet), but I just don't like leaving it in scope after it had its use very much either.
For me, many functions of Mudlet code are far too long with far too many variables lying around, doing nothing.
I come from a place where functions fit on one page and do one thing, they are easy to grasp and end quickly.
To new developers, Mudlet source can seem like a big ball of code, not sure what comes out when you pull somewhere.

Kebap committed:
Merge pull request #13 from SlySven/spellcheck_2_characters_undoPreviousMistake

Fix: undo mistake in prior PR

Kebap committed:
Merge branch 'development' into spellcheck_2_characters

SlySven reviewed (approved):
Should be good to go, finally, now...

:relieved:

SlySven commented:
> It's a matter of taste at this point, that is why I reluctantly agreed to your change.
> I agree, it does no harm (yet), but I just don't like leaving it in scope after it had its use very much either.
For me, many functions of Mudlet code are far too long with far too many variables lying around, doing nothing.
I come from a place where functions fit on one page and do one thing, they are easy to grasp and end quickly.
To new developers, Mudlet source can seem like a big ball of code, not sure what comes out when you pull somewhere.

That is a fair comment - but it has the converse feature that you have more levels of `{`...`}` wrapped code indentation in each function/method. If you consider the desire to remove similar `{`...`}` wrapped blocks of code considered to be extraneous (making the code harder to follow) caused by the "*else-after-return*" anti-pattern this makes me wonder whether we need to have a "proper" discussion about this at the next MMMMM. :grinning:

Kebap committed:
Drop TODO comment

Signed-off-by: Kebap <kebap_spam@gmx.net>

Kebap commented:
Danger asked me to drop the TODO comments

> Source file src/TCommandLine.cpp includes a TODO! Can you make the change right away? If no, better not to create a TODO - they just tend to hang around and never get fixed.

That relates to this section in TCommandLine.cpp

```cpp
    if (handle_profile) {
        // TODO: Make icons for these?
        // if (!qApp->testAttribute(Qt::AA_DontShowIconsInMenus)) {
        //    action_addWord = new QAction(QIcon(QPixmap(qsl(":/icons/dictionary-add-word.png"))), tr("Add to user dictionary"));
        //    action_removeWord = new QAction(QIcon(QPixmap(qsl(":/icons/dictionary-remove-word.png"))), tr("Remove from user dictionary"));
        // } else {
        action_addWord = new QAction(tr("Add to user dictionary"));
```

So I did. If you feel the need to keep this idea around, please open a new issue on github :bulb:
